### PR TITLE
Driving in traffic example

### PR DESCRIPTION
**Pull request by Adaickalavan**
This example illustrates the training of an ego agent to drive in traffic using DreamerV2 (https://github.com/danijar/dreamerv2) reinforcement-learning algorithm.

Ego agent earns rewards based on the distance travelled and is penalised for colliding with other vehicles and for going off-road.

### CODE REVIEW

**Comment by liu-allan** [comment]
I have noticed this error during training and evaluation:
image

I'm not sure where or what the PRC file is so if you could let me know, I can get some more information about the error

File: examples/driving_in_traffic/setup.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    long_description=long_description,
+    long_description_content_type="text/markdown",
+    version="0.1.0",
+    packages=find_packages(),
+    include_package_data=True,
+    zip_safe=True,
+    python_requires=">=3.7",
+    install_requires=[
+        "setuptools>=41.0.0,!=50.0",
+        "smarts[camera-obs] @ git+https://git@github.com/huawei-noah/SMARTS@develop",
```
Comment (Gamenot) [inline examples/driving_in_traffic/setup.py:21]: Interesting, I have not seen this syntax before. it looks useful.

File: CHANGELOG.md [hunk 2]
```diff
@@ -26,6 +26,7 @@ Copy and pasting the git commit messages is __NOT__ enough.
 - Added `smarts.core.utils.import_utils` to help with the dynamic import of modules.
 - Added `single_agent` env wrapper and unit test. The wrapper converts a single-agent SMARTS environment's step and reset output to be compliant with gym spaces.
 - Added `rgb_image` env wrapper and unit test. The wrapper filters SMARTS environment observation and returns only top-down RGB image as observation.
+- Added `driving_in_traffic` reinforcement learning example. An ego agent is trained using DreamerV2 to drive as far and as fast as possible in heavy traffic, without colliding or going off-road.
```
Comment (Gamenot) [inline CHANGELOG.md:29]: Could we consider this a benchmark?

File: examples/driving_in_traffic/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+ARG DEBIAN_FRONTEND=noninteractive
+
+# Install libraries
+RUN apt-get update --fix-missing && \
+    apt-get install -y \
+        software-properties-common && \
+    add-apt-repository -y ppa:deadsnakes/ppa && \
+    apt-get update && \
+    apt-get install -y \
+        ffmpeg \
```
Comment (Gamenot) [inline examples/driving_in_traffic/Dockerfile:12]: Why is `ffmpeg` needed? Envision?

Comment (Adaickalavan) [reply examples/driving_in_traffic/Dockerfile:12]: `ffmpeg` is needed to create videos from rgb images, to be displayed in tensorboard, to visualize training/evaluation progress.

File: CHANGELOG.md (hunk 2 above)
Comment (Adaickalavan) [reply CHANGELOG.md:29]: Yes, we could, if we want to.

File: examples/driving_in_traffic/setup.py (hunk 1 above)
Comment (Adaickalavan) [reply examples/driving_in_traffic/setup.py:21]: This needs to be replaced with a fixed smarts version, once we have a smarts release.

File: examples/driving_in_traffic/README.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+$ scl scenario build-all --clean ./scenarios/loop
+```
+
+## Train
+1. Train
+    ```bash
+    $ cd <path>/SMARTS/examples/driving_in_traffic
+    $ python3.7 run.py 
+    ```
+1. Trained model is saved into `<path>/SMARTS/examples/driving_in_traffic/logs/<folder_name>` folder.
```
Comment (Gamenot) [inline examples/driving_in_traffic/README.md:44]:
```suggestion
1. The trained model is saved into `<path>/SMARTS/examples/driving_in_traffic/logs/<folder_name>` folder.
```

File: examples/driving_in_traffic/README.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+1. Evaluate
+    ```bash
+    $ cd <path>/SMARTS/examples/driving_in_traffic
+    $ scl envision start -s ./scenarios &
+    $ python3.7 run.py --mode=evaluate --logdir="<path>/SMARTS/examples/driving_in_traffic/logs/<folder_name>" --head
+    ```
+1. Go to `localhost:8081` to view the simulation in Envision.
+
+## Docker
+1. Build and train inside docker container
```
Comment (Gamenot) [inline examples/driving_in_traffic/README.md:56]:
```suggestion
1. Build and train a DreamerV2 model inside docker container:
```

File: examples/driving_in_traffic/README.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
+# Driving in Traffic
+This example illustrates the training of an ego agent to drive, as fast and as far as possible, in traffic using DreamerV2 (https://github.com/danijar/dreamerv2) reinforcement-learning algorithm.
```
Comment (Gamenot) [inline examples/driving_in_traffic/README.md:2]:
```suggestion
This example illustrates the training of an ego agent to drive, as fast and as far as possible, in traffic using the DreamerV2 (https://github.com/danijar/dreamerv2) reinforcement-learning algorithm.
```

File: examples/driving_in_traffic/README.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+Ego agent earns rewards based on the distance travelled and is penalised for colliding with other vehicles and for going off-road.
```
Comment (Gamenot) [inline examples/driving_in_traffic/README.md:4]:
```suggestion
The ego agent earns rewards for the distance travelled and is penalised for colliding with other vehicles and for going off-road.
```

File: examples/driving_in_traffic/README.md (hunk 4 above)
Comment (Adaickalavan) [reply examples/driving_in_traffic/README.md]: Changed.

File: examples/driving_in_traffic/README.md (hunk 5 above)
Comment (Adaickalavan) [reply examples/driving_in_traffic/README.md]: Changed.

File: examples/driving_in_traffic/README.md (hunk 6 above)
Comment (Adaickalavan) [reply examples/driving_in_traffic/README.md]: Changed.

File: examples/driving_in_traffic/README.md (hunk 7 above)
Comment (Adaickalavan) [reply examples/driving_in_traffic/README.md]: Changed.

**Review by Gamenot — APPROVED** [review]
I think I have no issues with the content of the example although I think it would be better set out as a notebook rather than a standard python file.

**Review by RutvikGupta — APPROVED** [review]

**Comment by Adaickalavan** [comment]
> I think I have no issues with the content of the example although I think it would be better set out as a notebook rather than a standard python file.

Worry not, the example notebook is in development in another branch. I will merge it when it looks good.

**Pull request merged**